lib/winapi/profiling.c

```diff
@@ -11,14 +14,44 @@ BOOL QueryPerformanceCounter (LARGE_INTEGER *lpPerformanceCount)
 {
     assert(lpPerformanceCount != NULL);
 
-    lpPerformanceCount->QuadPart = KeQueryPerformanceCounter();
+    lpPerformanceCount->QuadPart = __rdtsc();
     return TRUE;
 }
 
 BOOL QueryPerformanceFrequency (LARGE_INTEGER *lpFrequency)
 {
     assert(lpFrequency != NULL);
 
-    lpFrequency->QuadPart = KeQueryPerformanceFrequency();
+#ifdef USE_RDTSC_FOR_FREQ
+    #define AVG_SET 10
+    ULARGE_INTEGER f_rdtsc, avg;
+    ULONG f_ticks = 0;
+
+    avg.QuadPart = 0;
+
+    for (int i = 0; i < AVG_SET; i++) {
+        ULARGE_INTEGER s_rdtsc;
+        ULONG s_ticks;
+
+        s_rdtsc.QuadPart = __rdtsc();
+        s_ticks = KeTickCount;
+
+        s_rdtsc.QuadPart -= f_rdtsc.QuadPart;
+        s_rdtsc.QuadPart /= s_ticks - f_ticks;
+
+        f_rdtsc.QuadPart = __rdtsc();
+        f_ticks = KeTickCount;
+
+        // Skip the first result as invalid
+        if (i)
+            avg.QuadPart += s_rdtsc.QuadPart;
+        
+        // If we call rdtsc too fast we'll end up with div by 0
+        Sleep(10);
```

Comment:
Correct it doesn't, that's added by me. It only ever replies 733mhz. I'd like to hear ideas of how to get this dynamically. Perhaps a run once and global var?

Reply:
I agree with jfr, code with this kind of performance issue isn't suitable for this type of functionality.
The loop doesn't make sense to me either - doing it in multiple small steps just accumulates rounding and measurement errors, exacerbated by using a low-precision timing source such as the tick counter.

We should calibrate the performance counter frequency value only once, so this function won't have to do anything more than just return the content of a global static variable.
The variable can be initialized via a function marked as a constructor to avoid having to do any manual initialization.
The timing source for the calibration should be `KeQueryPerformanceCounter`, which should give us better precision given the ACPI timer's tick rate of 3.579545 MHz. It may only support a 24 bit counter, so the code needs to handle a potential overflow during calibration.